### PR TITLE
Add SiameseDataset

### PR DESCRIPTION
**Pull request by yuyu2172**

### CODE REVIEW

**Comment by yuyu2172** [comment]
Updated

File: chainercv/datasets/siamese_dataset.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import numpy as np
+
+import chainer
+
+
+def _construct_label_to_key(labels):
+    d = dict()
+    for i, label in enumerate(labels):
+        if label not in d:
+            d[label] = list()
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: How about using `collections.defaultdict`?

Comment (yuyu2172) [reply chainercv/datasets/siamese_dataset.py]: Nice

File: docs/source/reference/datasets.rst [hunk 2]
```diff
@@ -14,6 +14,10 @@ directory_parsing_label_names
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 .. autofunction:: directory_parsing_label_names
 
+SiameseDataset
+~~~~~~~~~~~~~~
+.. autofunction:: SiameseDataset
```
Comment (Hakuyume) [inline docs/source/reference/datasets.rst]: `autoclass`?

Comment (yuyu2172) [reply docs/source/reference/datasets.rst]: Thanks

File: chainercv/datasets/siamese_dataset.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    Args:
+        dataset_0: The first base dataset.
+        dataset_1: The second base dataset.
+        pos_ratio (None or float): If this is not :obj:`None`,
+            this dataset tries to construct positive pairs at the
+            given rate. If :obj:`None`,
+            this dataset randomly samples examples from the base
+            datasets. The default value is :obj:`None`.
+        length (None): The length of this dataset. If :obj:`None`,
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: `length (int)`
I think we can omit default `None` form the type description. i.e. `pos_ratio (None or float)` -> `pos_ratio (float)`

File: chainercv/datasets/siamese_dataset.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            the length of the first base dataset is the length of this
+            dataset.
+        labels_0 (None or numpy.ndarray): The labels associated to
+            the first base dataset. The length should be the same as
+            the length of the first dataset. If this is :obj:`None`,
+            the labels are automatically fetched using the following
+            line of code: :obj:`[ex[1] for ex in dataset_0]`.
+            Also, if :obj:`pos_ratio` is :obj:`None`, this value
+            is ignored. The default value is :obj:`None`.
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: How about describing the intention of this argument?
```
...
the labels are automatically fetched using the following line of code: :obj:`[ex[1] for ex in dataset_0]`.
If an array of labels already exists, users can skip this fetching iteration and reduce computation cost.
...
```

File: chainercv/datasets/siamese_dataset.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        self._pos_ratio = pos_ratio
+        if length is None:
+            length = len(self._dataset_0)
+        self._length = length
+
+        if pos_ratio is not None:
+            if labels_0 is None:
+                labels_0 = [example[1] for example in dataset_0]
+            if labels_1 is None:
+                labels_1 = [example[1] for example in dataset_1]
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: If `dataset_1 is dataset_0`, we can use `labels_0` as `labels1`. I think using same dataset for both `dataset_0` and `dataset_1` is very common and this optimization is effective.

File: chainercv/datasets/siamese_dataset.py (hunk 4 above)
Comment (Hakuyume) [reply chainercv/datasets/siamese_dataset.py]: Even if `labels_0` and `labels_1` are specified, `_construct_label_to_key` iterates these arrays. So, I think the effect of this argument is limited. Is this argument necessary?

Comment (yuyu2172) [reply chainercv/datasets/siamese_dataset.py]: `_construct_label_to_key` would not iterate over the dataset, so no images would be loaded if `labels_*` are specified.

Comment (Hakuyume) [reply chainercv/datasets/siamese_dataset.py]: I see.

File: chainercv/datasets/siamese_dataset.py [hunk 6]
```diff
@@ -57,6 +61,8 @@ def __init__(self, dataset_0, dataset_1,
         self._length = length
 
         if pos_ratio is not None:
+            if labels_1 is None and  dataset_0 is dataset_1:
+                labels_1 = labels_0
             if labels_0 is None:
                 labels_0 = [example[1] for example in dataset_0]
             if labels_1 is None:
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: In the case of `labels_0 is None and labels_1 is None`, we can use `[example[1] for example in dataset_0]` for `lablels_1`. Also, the computation of `self._label_to_index_1` can be skipped.

File: chainercv/datasets/siamese_dataset.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+        if pos_ratio is not None:
+            if labels_1 is None and dataset_0 is dataset_1:
+                labels_1 = labels_0
+            if labels_0 is None and dataset_0 is dataset_1:
+                labels_0 = labels_1
+            if labels_0 is None:
+                labels_0 = [example[1] for example in dataset_0]
+            if labels_1 is None:
+                labels_1 = [example[1] for example in dataset_1]
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: In the case of `labels_0 is None`, `labels_1 is None` and `dataset_0 is dataset_1`,
```python
        if pos_ratio is not None:
            if labels_1 is None and dataset_0 is dataset_1:
                labels_1 = labels_0  # labels_1 = None
            if labels_0 is None and dataset_0 is dataset_1:
                labels_0 = labels_1  # labels_0 = None
            if labels_0 is None:
                labels_0 = [example[1] for example in dataset_0]  # fetch labels from dataset_0 <- OK
            if labels_1 is None:
                labels_1 = [example[1] for example in dataset_1]  # fetch labels from dataset_1 <- redundant
```

File: chainercv/datasets/siamese_dataset.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            labels_1 = np.array(labels_1)
+            if not (labels_0.dtype == np.int32 and labels_0.ndim == 1
+                    and len(labels_0) == len(dataset_0) and
+                    labels_1.dtype == np.int32 and labels_1.ndim == 1
+                    and len(labels_1) == len(dataset_1)):
+                raise ValueError('the labels are invalid.')
+
+            # Construct mapping label->idx
+            self._label_to_index_0 = _construct_label_to_key(labels_0)
+            self._label_to_index_1 = _construct_label_to_key(labels_1)
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]:
```python
if dataset_0 is dataset_1:
    self,_label_to_index_1 = self._label_to_index0
else:
    self._label_to_index_1 = _construct_label_to_key(labels_1)
```

File: chainercv/datasets/siamese_dataset.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                elif labels_1 is None:
+                    labels_1 = labels_0
+            else:
+                if labels_0 is None:
+                    labels_0 = [example[1] for example in dataset_0]
+                if labels_1 is None:
+                    labels_1 = [example[1] for example in dataset_1]
+
+            labels_0 = np.array(labels_0)
+            labels_1 = np.array(labels_1)
```
Comment (Hakuyume) [inline chainercv/datasets/siamese_dataset.py]: The following code can reduce call of `np.array`.
```python
            if dataset_0 is dataset_1:
                if labels_0 is None and labels_1 is None:
                    labels_0 = np.array([example[1] for example in dataset_0])
                    labels_1 = labels_0
                elif labels_0 is None:
                    labels_0 = labels_1
                elif labels_1 is None:
                    labels_1 = labels_0
            else:
                if labels_0 is None:
                    labels_0 = np.array([example[1] for example in dataset_0])
                if labels_1 is None:
                    labels_1 = np.array([example[1] for example in dataset_1])
```

**Pull request merged**